### ISSUE TITLE
Conditional forwarding using MAC not working

### ISSUE THREAD

**Issue by mojo333** [boostchicken/udm-utilities#5]
If i make changes to` /mnt/data/nextdns/nextdns.conf`, in order for these changes to take effect do I just need to stop and start the nextdns container i.e.
```
podman stop nextdns
podman start nextdns
```

I have made a configuration change and it is not taking effect and I want to rule out the container before raising the issue with nextdns. I am not familiar with podman.

**Comment by mojo333** [comment]
@sfhoo on_boot.sh is the hook for enabling nextdns on reboot and also has the DNAT rules to ensure that all DNS traffic on WAN out (br0) is redirected to nextdns if needed - at least my understanding.

I checked that the conditional forwarding pattern matching with my config by running my config (above) through NextDNS test case for this: `https://github.com/nextdns/nextdns/blob/master/config/id_test.go` and it correctly selects the right configuration based on either IP or MAC address so it's definitely not a config issue.

I also validated that podman stop / start works as expected in that the new nextdns.conf is picked up when changes (did this by enabling / disabling logging - which was enabled/disabled on stop/start).

The only thing I can think of is that mDNS isn't working as expected in that the device MAC address isn't being picked up for some reason (but the IP address is always picked up).

So in short, any conditional forwarding by IP address works perfectly across all my VLANs; any conditional forwarding by MAC address does not work for me - can anyone confirm this?

**Comment by boostchicken** [comment]
@mojo333 have you tried making sure your bridge is in promiscuous mode?

**Comment by mojo333** [comment]
@boostchicken i saw the update to on_boot.sh and applied it. Rebooted my UDM PRO and checked  that it's enabled, but no luck, still the same outcome as above i.e. conditional filtering on IP works perfectly but by MAC is not working at all. 

In my nextdns.conf I have ensured that the MAC conditional forward is the 1st item in the list as according to nextdns it's first hit wins, but still no luck.

Is this something you could do a quick test on your side i.e. create a new profile in nextdns and try to push one of your devices to it using it's MAC address?

**Comment by boostchicken** [comment]
@mojo333 
would it be possible to do a tcpdump and see what's coming through in the container and at the udm?  The mac address SHOULD be forwarded with the request, my guess is it's getting overwritten with the bridges mac address.

**Comment by mojo333** [comment]
@boostchicken I suspect you're spot on with this guess but let me look into this.

**Comment by boostchicken** [comment]
@mojo333 , it is a little bit out of my skillset as well.  The reason we have to make this bridge  is because macvlan is designed to limit communication between the host and its children.  By making this bridge interface we allow anything to talk to it by going through the br5.mac interface.  So we may be stuck with this.

I wanted to do ipvlan but Unifi didn't add that to the kernel the put on the box.  

All that to say, I am not sure this is fixable.  I would suggest maybe forwarding on a CIDR as opposed to MAC?  I will spend some time to see if there is any possible way to fix this.

**Comment by boostchicken** [comment]
@mojo333 some quick reading showed me 
```
When you create a macvlan network, it can either be in bridge mode or 802.1q trunk bridge mode.

In bridge mode, macvlan traffic goes through a physical device on the host.

In 802.1q trunk bridge mode, traffic goes through an 802.1q sub-interface which Docker creates on the fly. This allows you to control routing and filtering at a more granular level.```

https://docs.docker.com/network/macvlan/

So in [20-dns.conflist](https://github.com/boostchicken/udm-utilities/blob/master/nextdns/udm-files/20-dns.conflist) you will notice the mode is set to "bridge".  The other possible values are documented [here](https://developers.redhat.com/blog/2018/10/22/introduction-to-linux-interfaces-for-virtual-networking/#macvlan)

My guess is the mode you are looking for is "source" or "passthru".  Give them a shot and lot me know.

Thanks!

**Comment by boostchicken** [comment]
Actually come to think of it, wouldn't the incoming mac be the one the matters? Wouldn't the nextdns client handle that conditional forwarding at the proxy?

**Comment by mojo333** [comment]
@boostchicken thanks for your thoughts on this - good avenues to investigate. Running two containers would certainly be one solution. On your last point, I had the same thought and I need to look into this more as it makes sense that the request from the nextdns client would contain the identifier - so done at source not destination. Might need to fork nextdns and add logging of the mac address but then i'd need to learn go!

**Comment by boostchicken** [comment]
@mojo333 one thing that might solve this is not having it on a different vlan.  Anyway you slice it moving from vlan to vlan is going to require some routing.  I don't know how complex your setup is but I ran a quick test.  I moved it to my br0 and everything looked good for incoming traffic.

The reason for this is it's going to have to hit the br5 gateway always and your mac will be lost, it is never a direct connect.  With it on the same vlan, you can actually hit the container directly with no other interfaces in the way.  Only egress has to go through the br0.mac interface to defeat the linux kernel filtering settings.
